### PR TITLE
boot/grub.go: add new bootchain

### PR DESCRIPTION
**Pull request by valentindavid**
We need to resolve the boot chains another place based on the trusted assets we encountered to be installed. At this point it could be any chain. We will need to discover later what the correct chain is.

Also make TrustedAssets return an unsorted data structure to make sure we do not use the order like the comments claimed.

### CODE REVIEW

**Comment by valentindavid** [comment]
Needed for #13205

File: boot/assets.go
Comment (valentindavid) [inline boot/assets.go:260]: I have changed this type to be able to see all the place were we could have expected order to have meaning. This change can be reverted before merge.

File: boot/seal.go
Comment (valentindavid) [inline boot/seal.go:755]: Reminder: this is missing a test

File: boot/seal.go
Comment (valentindavid) [inline boot/seal.go:830]: Reminder, this is missing a test.

**Comment by valentindavid** [comment]
I have found a bug. We do not seem to remove files that should be removed. So moving from boot entry back to non boot entry (that is removing the `fbx64.efi`)  would not work.

I wonder if we should add "removed" content to the gadget.yaml.

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:810]: trustedAssets is not sorted and this code remembers only one asset, isn't there a chance that we will find exactly the one that we didn't remember?

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:827]: are there new tests for the case of multiple asset for a single name? per the comment above I'm not quite sure whether this works sensibly  or not? could we instead consider to have some control here or in the bootloaader so that we store the new clashing assets under a different name?

EFI/boot/grubx64.go  => assetName = grubx64.go
EFI/ubuntu/grubx64.go => assetName  = ubuntu:grubx64.go

or similar, perhaps?

**Review by pedronis — COMMENTED** [review]
did a pass, I'm slightly wary of the changes to assets.go

File: bootloader/bootloader.go
Comment (pedronis) [inline bootloader/bootloader.go:225]: this could be changed to be a map:

relativePath => assetName

for example

Comment (valentindavid) [reply bootloader/bootloader.go:225]: I wonder if we could use that to disambiguate conflicting base names.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go:827]: I will try that.

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go:830]: I think we are missing a lot of tests for `runModeBootChain`. I will finish that when I am back.

**Comment by valentindavid** [comment]
I had to rebase. So I cleaned up the history.

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go:830]: Done.

File: tests/nested/manual/uc-grub-boot-chains/modify-gadget.py
Comment (pedronis) [inline tests/nested/manual/uc-grub-boot-chains/modify-gadget.py:5]: edition_bump perhaps?

also I think this argument would be clearer as an option instead of a positional arg

File: tests/nested/manual/uc-grub-boot-chains/modify-gadget.py
Comment (pedronis) [inline tests/nested/manual/uc-grub-boot-chains/modify-gadget.py:6]: this could use a comment to explain the two choices

File: tests/nested/manual/uc-grub-boot-chains/task.yaml
Comment (pedronis) [inline tests/nested/manual/uc-grub-boot-chains/task.yaml:95]: is this not supported? will it be?

File: bootloader/grub.go
Comment (pedronis) [inline bootloader/grub.go:642]: this is a bit strange given that BootFile has a still a field called Path, I thought we would keep the path as they were and use the TrustedAssets map to get to id to use in the cache

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:798]: why the []string?  assetNames are noew unique again and have one asset assigned to them?

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:729]: buildBootAssets could take the TrustedAssets map and then BootFile could still carry paths?

File: bootloader/grub.go
Comment (pedronis) [inline bootloader/grub.go:504]: nitpick: per go naming conventions this should just be Id

**Review by pedronis — COMMENTED** [review]
thank you, mainly some comments/questions about using Ids as Paths in BootFiles

File: tests/nested/manual/uc-grub-boot-chains/task.yaml
Comment (valentindavid) [reply tests/nested/manual/uc-grub-boot-chains/task.yaml:95]: Oh! I am missing a fixme here.

I need to fix the deletion of gadget content that is not available in the next gadget. In case of downgrade, the "next gadget"  should not have "fbx64.efi", so we should remove it, because it changes the behavior of shim.

I will do this fix in a different PR. But I will for sure comment here on why this does not work. And why it needs to work.

File: tests/nested/manual/uc-grub-boot-chains/modify-gadget.py
Comment (valentindavid) [reply tests/nested/manual/uc-grub-boot-chains/modify-gadget.py:14]: Done.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go]: I was verifying that we had only one later. But now I have rewritten it so that we verify it right away.

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go]: Done

File: bootloader/grub.go
Comment (valentindavid) [reply bootloader/grub.go]: Done.

File: tests/nested/manual/uc-grub-boot-chains/modify-gadget.py
Comment (valentindavid) [reply tests/nested/manual/uc-grub-boot-chains/modify-gadget.py]: Done.

File: bootloader/grub.go
Comment (valentindavid) [reply bootloader/grub.go:642]: Thank you for the tip. I did not realize I could do that. I have fixed it.

**Comment by valentindavid** [comment]
@olivercalder I have added you as reviewer. This has some changes from your branch to add the new boot chain. But it does not do anything with the variables yet. I will rebase your branch on top of this. I thought you would like to have a look.

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:260]: probably a good idea to add a quick doc comment about what is the key and value of this

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:264]: same

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:429]: same

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:434]: same,  another approach is to define a type for this and explain the two side in the type definition

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:804]: we probably want to keep this comment before the log below?

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:879]: this is strange, it should either be a log or a panic, no?

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:675]: same, either log or panic

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:789]: same

**Review by pedronis — COMMENTED** [review]
thank you, some detail comments on the last round of changes

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go:804]: Not all trusted assets will be present now. So I am thinking we should remove the notice. But the logic is still fine. If the asset is not in the cache, then we should ignore it.

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go]: I removed it.

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go]: I removed it.

File: boot/seal_test.go
Comment (pedronis) [reply boot/seal_test.go:1259]: I see, do have higher level tests then that test the case where we need to consider both old and new grub assets locations?  I saw only this one doing that but maybe I missed them

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:873]: this probably merit a comment why the new logic is ok with not finding some asset chains

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:882]: same

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:817]: and a comment here

**Review by pedronis — COMMENTED** [review]
thank you, looking good, just some remarks about comments and a question about the test combinations

File: boot/seal.go
Comment (valentindavid) [reply boot/seal.go:873]: This one can only be generated by a broken bootloader. So it should be an error.

File: boot/seal_test.go
Comment (valentindavid) [reply boot/seal_test.go:1259]: Right, I think we are missing a test for `ResealKeyToModeenv` with both chain present at the same time. We do test both. But separately.

**Review by alfonsosanchezbeato — COMMENTED** [review]
I have a question about the spread test

File: tests/nested/manual/uc-grub-boot-chains/task.yaml
Comment (valentindavid) [reply tests/nested/manual/uc-grub-boot-chains/task.yaml:92]: Well, we do not touch variables in this PR. This will come in the Oliver's PR. It is better to remove `EFI/ubuntu` and `EFI/boot/fbx64.efi`. If the variables are wrong, it will fallback anyway. But this is what I want installation of gadget to do. A "content" entry that disappears should imply removing the file.

File: boot/seal_test.go
Comment (valentindavid) [reply boot/seal_test.go:1442]: I have modified `TestResealKeyToModeenvWithSystemFallback` to also cover cases where we upgrade or downgrade. This calls `ResealKeyToModeenv` which is higher level.

But that is starting to be complicated. Specially because of ordering and factorization of boot chains. There is lots of ways to express allowed boot chains. But we only test one, which is very hard to guess. I spend a day to make it work.

File: boot/seal.go
Comment (pedronis) [inline boot/seal.go:881]: s/Asset/internal error: asset/   error always start with lowercase

**Review by pedronis — COMMENTED** [review]
thank you, small issue with the error message also there seem to be formatting issues as the static checks are failing

File: tests/nested/manual/uc-grub-boot-chains/task.yaml
Comment (alfonsosanchezbeato) [reply tests/nested/manual/uc-grub-boot-chains/task.yaml:92]: But maybe still worth running this part of the test by removing "manually" the files, even while keeping the comment so when removing from snapd is implemented we can remove the workaround?

File: bootloader/grub_test.go
Comment (olivercalder) [inline bootloader/grub_test.go:4]: 2024 here as well :)

**Review by olivercalder — COMMENTED** [review]
Looks really good, thank you!

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go:817]: I changed this multiple times. So it seems in the end it was just about filtering duplicates. I will simplify the code here.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go:824]: That depends on the role though.

Comment (valentindavid) [reply boot/assets.go:824]:
```
	trustedAssetsMap := &m.CurrentTrustedBootAssets
[...]
		trustedAssetsMap = &m.CurrentTrustedRecoveryBootAssets
```

File: bootloader/grub.go
Comment (valentindavid) [reply bootloader/grub.go:533]: Yes, I should may comment on it. The idea about using "asset names" instead of the path is because current implementation uses base names. So for backward compliance, we need to use no tag for all the paths that were used before. I will comment it.

Comment (valentindavid) [reply bootloader/grub.go:533]: Ah now I understand. fb was not an old one. I can put boot: for this one.

File: bootloader/bootloadertest/bootloadertest.go
Comment (valentindavid) [reply bootloader/bootloadertest/bootloadertest.go]: Changed it and all the tests to use a map.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go]: I have simplified the code.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go:807]: I have put a comment that I think is easier to understand now.

File: boot/assets.go
Comment (valentindavid) [reply boot/assets.go]: Correct. Not easy to spot. I have removed it.

File: tests/nested/manual/uc-grub-boot-chains/task.yaml
Comment (valentindavid) [reply tests/nested/manual/uc-grub-boot-chains/task.yaml:92]: I will try preventing reboot with systemd-inhibit and see if I can remove the assets before reboot.

Comment (alfonsosanchezbeato) [reply tests/nested/manual/uc-grub-boot-chains/task.yaml:92]: Right, it is true that being faster than the reboot can be a problem, so please ignore this if it gets too hacky. Approving now, thanks for the changes.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM

File: boot/assets.go
Comment (pedronis) [inline boot/assets.go:818]: I wonder as we are very much in control of this from the bootloader side, whether this should be an actual error?  cc @alfonsosanchezbeato @bboozzoo

**Review by pedronis — APPROVED** [review]
thank you, one wondering

File: boot/assets.go
Comment (alfonsosanchezbeato) [reply boot/assets.go:818]: Probably this should return as an internal error, I agree

**Review by olivercalder — APPROVED** [review]
Looks great, thanks for taking on all of this!

**Comment by valentindavid** [comment]
Everything should be fixed. I will let the CI run and then autosquash it.

**Pull request merged**
